In [1]:
import numpy as np
import pandas as pd
import re
pd.set_option('max_rows',200)

In [2]:
accred_df = pd.read_csv('Accreditation_Data.csv')
sample_df = pd.read_csv('sampleTestFile.csv')
test_df = pd.read_csv('testFile.csv')

dfs = [accred_df, sample_df, test_df]

print(len(accred_df))
print(len(sample_df))
print(len(test_df))



4383
100
392


Next step - treatment of URLs

In [3]:
for df in dfs:
    df['URL'] = df['URL'].str.lower()

def strip_url(url):
    stripped_url = url.replace('http://','').replace('https://','').replace('www.','')
    match = re.match("(.*)\.",stripped_url)
    
    if match:
        print(match.groups()[0])
        return match.groups()[0]
    else:
        print('nomatch:',stripped_url)
        return stripped_url
    
for df in dfs:
    df['stripped_url'] = df['URL'].apply(strip_url)

acc-careers
acc-careers
accschool.peedeeworld
acmt
aesa
aihs
akronbeautyschool
akronschools
alexacebeauty
alliedteched
americanbeautyacademy
americancollegeofhair
antiochcollege
antonellic
aoma
artisticbeautycolleges
artisticbeautycolleges
ashdowncollege
ati.ag.ohio-state
aticareertraining
audioschool
avtec.labor.state.ak
awc
bc.inter
beacon
bealcollege
beautycareers
beautyschool
bellevue
bellin
bellinghambeautyschool
beonair
berkeley.peralta
beta.nwiht
blottsalonschools
blue.ab
boe.kana.k12.wv
boe.mono.k12.wv
boe.putn.k12.wv
bramsonort
brc
brewstertech
brioacademy
brownsontechnicalschool
bryman-college
bryman-college
burlingtontech
californiacareerschool
canadacollege
capitol-college
carnegieinstitute
carouselbeauty
carouselbeauty
carouselbeauty
cccua
cci
cci
cci
ccsce
cdtschool
ceitraining
cempr
centralcareer
centralohio.dalecarnegie
century-school.bizhosting
cetcleveland
cfinstitute
charlesstuartschool
chase
nomatch: clank85462
classact1cosmetology
coastline.cccd
collegeofhairdesign

In [4]:
## this function does several things at once:
## 1) we now also split on hyphens. Webpages may be hyphenated, but this way we'll still pick them up instead of missing
## them if they drop the hyphen in URL.
## 2) apostrophes aren't allowed in URLs, so we delete all apostrophes.
## 3) any generic words like College, School etc. are omitted, because they are not specific enough to lead to a match.

def get_keywords(x):
    possible_words = re.split(r'\s|-', x) ## split on either space or hyphen
    omit_words = ['college','institute','school','schools','university','the','of',',','-','and']
    possible_words = [ word.lower() for word in possible_words ]  ## convert everything to lower case
    resultwords  = [word.replace('\'', '') for word in possible_words if word.lower() not in omit_words]  
    ## remove any apostrophes from words
    resultwords = list(filter(None, resultwords)) # fastest way to remove any new empty strings
    return resultwords

for df in dfs:
    df['keywords'] = df['Institution'].apply(get_keywords)

next: many websites are acronyms of the institution name. let's figure out an acronym for each name.

In [5]:
#following function returns potential college acronym in lowercase
#one quick tweak from staring at data - some places start with The (uppercase T); let's 
def acronym(s):
    s = s.replace('The ','') #specifically handles initial The (which tends to mess up acronyms)
    return "".join(c.lower() for c in s if c.isupper())

for df in dfs:
    df['acronym'] = df['Institution'].map(acronym)

In [6]:
def make_acronyms(X):
    
    if len(X['acronym']) >= 3:  ## length limit because one-letter acronyms risk false matches...
        return list([X['acronym'], X['acronym'][1:], X['acronym'][:-1]])
    else:
        return [X['acronym']]

for df in dfs:
    df['potential_acronyms'] = df.apply(make_acronyms, axis=1)
    
## list comprehension that cycles through all potential acronyms and checks if any are in url.
def check_any_acronym(x):
    in_url = ([ ac in x['stripped_url'] for ac in x['potential_acronyms'] ])
    return(any(in_url))

for df in dfs:
    df['potential_acronym_in_url'] = df.apply(check_any_acronym, axis=1)

In [7]:
accred_df

Institution  \
0                  American Commercial College of Texas   
1                 American Commercial College - Abilene   
2                          Anson College of Cosmetology   
3                American College of Medical Technology   
4     Aviation and Electronic Schools of America - C...   
5                American University of Health Sciences   
6                            Gerber Akron Beauty School   
7                             Adult Vocational Services   
8                  Alexandria Academy of Beauty Culture   
9                           Fortis Institute - Scranton   
10                    American Beauty Academy - Wheaton   
11       American College of Hairstyling - Cedar Rapids   
12                                      Antioch College   
13                      Antonelli College - Hattiesburg   
14               Academy of Oriental Medicine at Austin   
15                      Empire Beauty School - Glendale   
16                   Empire Beauty School - Westminster   
17                   Ashdown College of Health Sciences   
18    Ohio State University Agricultural Technical I...   
19                   ATI Career Training Center - Miami   
20                          Institute of Audio Research   
21                   Alaska Vocational Technical Center   
22                           Allegheny Wesleyan College   
23     Inter American University of Puerto Rico Bayamon   
24                                    Beacon University   
25                                         Beal College   
26                             Career Academy of Beauty   
27                        Career Academy of Hair Design   
28                                  Bellevue University   
29    Bellin College/ Bellin Health School of Radiol...   
30                             Bellingham Beauty School   
31       Ohio Center for Broadcasting - Colorado Campus   
32                                Berkeley City College   
33      Northwestern Institute of Health and Technology   
34                     Ohio State School of Cosmetology   
35                         Alderson Broaddus University   
36     Ben Franklin Career & Technical Education Center   
37         Monongalia County Technical Education Center   
38                   Putnam Career and Technical Center   
39                                  Bramson ORT College   
40                                  Baton Rouge College   
41                            Brewster Technical Center   
42                           Marinello School of Beauty   
43                            Brownson Technical School   
44                        Everest College - Los Angeles   
45                           Everest College - Torrance   
46                          Burlington Technical Center   
47                             California Career School   
48                                       Canada College   
49                        Capitol Technology University   
50                              Carnegie Career College   
51                              Carousel Beauty College   
52              Carousel of Miami Valley Beauty College   
53          Carousel Beauty College & The Spa Institute   
54    Cossatot Community College of the University o...   
55                         Everest Institute - San Jose   
56                         Everest Institute - Brighton   
57                          Everest Institute - Hayward   
58    Central California School of Continuing Educat...   
59                           Commercial Driver Training   
60                 Career Education Institute-Henderson   
61               Centro de Estudios Multidisciplinarios   
62                                Central Career School   
63               Dale Carnegie Training of Central Ohio   
64        Ohio Academy - A Paul Mitchell Partner School   
65           Center for Employment Training - Cleveland   
66              Central Florida Institute - Palm Harbor   
67                              

In [8]:
## testing out a different version of make_acronyms that tries out a couple of other possible acronyms - will see if
## increases probability of false match too much.

def make_acronyms_extra(X):
    
    ## one initial acronym try - the first letter of every word, even filler words like of, the etc.
    ## low success rate, but does happen sometimes and no other way of catching such cases
    acronym_allwords = "".join(word[0].lower() for word in X['Institution'].split())
    
    if len(X['acronym']) >= 3:  ## length limit because one-letter acronyms risk false matches...
        
        acs = list([X['acronym'], X['acronym'][1:], X['acronym'][:-1]])
        
        acs.append(acronym_allwords.replace('-',''))
        
        if len(X['acronym']) >= 4:
            acs.append(X['acronym'][:-2])
        
        print(acs)
        return acs
    
    else:
        acs = [X['acronym'],acronym_allwords]
        print(acs)
        return acs

for df in dfs:
    df['potential_acronyms_extra'] = df.apply(make_acronyms_extra, axis=1)
    
## list comprehension that cycles through all potential acronyms and checks if any are in url.
def check_any_acronym_extra(x):
    in_url = ([ ac in x['stripped_url'] for ac in x['potential_acronyms_extra'] ])
    return(any(in_url))

for df in dfs:
    df['potential_acronym_in_url_extra'] = df.apply(check_any_acronym_extra, axis=1)

['acct', 'cct', 'acc', 'accot', 'ac']
['acca', 'cca', 'acc', 'acca', 'ac']
['acc', 'cc', 'ac', 'acoc']
['acmt', 'cmt', 'acm', 'acomt', 'ac']
['aesac', 'esac', 'aesa', 'aaesoac', 'aes']
['auhs', 'uhs', 'auh', 'auohs', 'au']
['gabs', 'abs', 'gab', 'gabs', 'ga']
['avs', 'vs', 'av', 'avs']
['aabc', 'abc', 'aab', 'aaobc', 'aa']
['fis', 'is', 'fi', 'fis']
['abaw', 'baw', 'aba', 'abaw', 'ab']
['achcr', 'chcr', 'achc', 'acohcr', 'ach']
['ac', 'ac']
['ach', 'ch', 'ac', 'ach']
['aoma', 'oma', 'aom', 'aoomaa', 'ao']
['ebsg', 'bsg', 'ebs', 'ebsg', 'eb']
['ebsw', 'bsw', 'ebs', 'ebsw', 'eb']
['achs', 'chs', 'ach', 'acohs', 'ac']
['osuati', 'suati', 'osuat', 'osuati', 'osua']
['atictcm', 'tictcm', 'atictc', 'actcm', 'atict']
['iar', 'ar', 'ia', 'ioar']
['avtc', 'vtc', 'avt', 'avtc', 'av']
['awc', 'wc', 'aw', 'awc']
['iauprb', 'auprb', 'iaupr', 'iauoprb', 'iaup']
['bu', 'bu']
['bc', 'bc']
['cab', 'ab', 'ca', 'caob']
['cahd', 'ahd', 'cah', 'caohd', 'ca']
['bu', 'bu']
['bcbhsrt', 'cbhsrt', 'bcbhsr', 'bc

In [9]:
sample_df['correct_prediction_1'] = sample_df['isWrong'] == ~sample_df['potential_acronym_in_url']
sample_df['correct_prediction_2'] = sample_df['isWrong'] == ~sample_df['potential_acronym_in_url_extra']

print(sample_df['correct_prediction_1'].value_counts())
print(sample_df['correct_prediction_2'].value_counts())

True     67
False    33
Name: correct_prediction_1, dtype: int64
True     68
False    32
Name: correct_prediction_2, dtype: int64


In [10]:
sample_df[~sample_df['correct_prediction_1']]

Institution  \
3                           Suburban Technical School   
7                                       Baker College   
9   Baruch College of the City University of New York   
10                       Bay de Noc Community College   
13                             Benedictine University   
14                                 SUNY at Binghamton   
19               SUNY College of Technology at Canton   
21  New York City College of Technology of the Cit...   
23  SUNY College of Agriculture and Technology at ...   
29                        Sacramento Job Corps Center   
31                               The Juilliard School   
33                  The Master's College and Seminary   
34                         Seton Montessori Institute   
40                         Newport Business Institute   
42      New Mexico Institute of Mining and Technology   
43          Northampton County Area Community College   
49                       SUNY College at Old Westbury   
50                        SUNY College at Plattsburgh   
52                                 Radford University   
53                     Randolph-Macon Woman's College   
54                     Rappahannock Community College   
58                   The General Theological Seminary   
59                      Remington College - San Diego   
61                                    Rob Roy Academy   
62                          Rob Roy Academy - Taunton   
65                               Saint Paul's College   
68                                   Sunstate Academy   
70                              School of Visual Arts   
74              Northern Baptist Theological Seminary   
79                                Springfield College   
84            Saint Patrick's Seminary and University   
94            Teachers College of Columbia University   
99       The Restaurant School at Walnut Hill College   

                           URL  isWrong        stripped_url  \
3             suburbantech.com        0        suburbantech   
7                www.baker.edu        0               baker   
9          www.baruch.cuny.edu        0         baruch.cuny   
10          www.baycollege.edu        0          baycollege   
13                 www.ben.edu        0                 ben   
14          www.binghamton.edu        0          binghamton   
19              www.canton.edu        0              canton   
21       www.citytech.cuny.edu        0       citytech.cuny   
23         www.cobleskill.edu/        0          cobleskill   
29       www.jcdc.jobcorps.org        0       jcdc.jobcorps   
31           www.juilliard.edu        0           juilliard   
33             www.masters.edu        0             masters   
34       www.montessorieca.org        0       montessorieca   
40     www.newportbusiness.com        0     newportbusiness   
42                 www.nmt.edu        0                 nmt   
43         www.northampton.edu        0         northampton   
49         www.oldwestbury.edu        0         oldwestbury   
50         www.plattsburgh.edu        0         plattsburgh   
52             www.radford.edu        0             radford   
53     www.randolphcollege.edu        0     randolphcollege   
54        www.rappahannock.edu        0        rappahannock   
58    www.remingtoncollege.edu        1    remingtoncollege   
59   www.remingtoncollege.edu/        0    remingtoncollege   
61             www.rob-roy.com        0             rob-roy   
62             www.rob-roy.com        0             rob-roy   
65          www.saintpauls.edu        0          saintpauls   
68    www.sandiegojobcorps.org        1    sandiegojobcorps   
70  www.schoolofvisualarts.edu        0  schoolofvisualarts   
74            www.seminary.edu        0            seminary   
79  www.springfieldcollege.edu        0  springfieldcollege   
84  www.stpatricksseminary.org        0  stpatricksseminary   
94                  www.tc.edu        0                  tc   
99   www.walnuthillcolle

In [11]:
accred_df

Institution  \
0                  American Commercial College of Texas   
1                 American Commercial College - Abilene   
2                          Anson College of Cosmetology   
3                American College of Medical Technology   
4     Aviation and Electronic Schools of America - C...   
5                American University of Health Sciences   
6                            Gerber Akron Beauty School   
7                             Adult Vocational Services   
8                  Alexandria Academy of Beauty Culture   
9                           Fortis Institute - Scranton   
10                    American Beauty Academy - Wheaton   
11       American College of Hairstyling - Cedar Rapids   
12                                      Antioch College   
13                      Antonelli College - Hattiesburg   
14               Academy of Oriental Medicine at Austin   
15                      Empire Beauty School - Glendale   
16                   Empire Beauty School - Westminster   
17                   Ashdown College of Health Sciences   
18    Ohio State University Agricultural Technical I...   
19                   ATI Career Training Center - Miami   
20                          Institute of Audio Research   
21                   Alaska Vocational Technical Center   
22                           Allegheny Wesleyan College   
23     Inter American University of Puerto Rico Bayamon   
24                                    Beacon University   
25                                         Beal College   
26                             Career Academy of Beauty   
27                        Career Academy of Hair Design   
28                                  Bellevue University   
29    Bellin College/ Bellin Health School of Radiol...   
30                             Bellingham Beauty School   
31       Ohio Center for Broadcasting - Colorado Campus   
32                                Berkeley City College   
33      Northwestern Institute of Health and Technology   
34                     Ohio State School of Cosmetology   
35                         Alderson Broaddus University   
36     Ben Franklin Career & Technical Education Center   
37         Monongalia County Technical Education Center   
38                   Putnam Career and Technical Center   
39                                  Bramson ORT College   
40                                  Baton Rouge College   
41                            Brewster Technical Center   
42                           Marinello School of Beauty   
43                            Brownson Technical School   
44                        Everest College - Los Angeles   
45                           Everest College - Torrance   
46                          Burlington Technical Center   
47                             California Career School   
48                                       Canada College   
49                        Capitol Technology University   
50                              Carnegie Career College   
51                              Carousel Beauty College   
52              Carousel of Miami Valley Beauty College   
53          Carousel Beauty College & The Spa Institute   
54    Cossatot Community College of the University o...   
55                         Everest Institute - San Jose   
56                         Everest Institute - Brighton   
57                          Everest Institute - Hayward   
58    Central California School of Continuing Educat...   
59                           Commercial Driver Training   
60                 Career Education Institute-Henderson   
61               Centro de Estudios Multidisciplinarios   
62                                Central Career School   
63               Dale Carnegie Training of Central Ohio   
64        Ohio Academy - A Paul Mitchell Partner School   
65           Center for Employment Training - Cleveland   
66              Central Florida Institute - Palm Harbor   
67                              

In [12]:
print(accred_df['potential_acronym_in_url'].value_counts(normalize = True))
print(accred_df['potential_acronym_in_url_extra'].value_counts(normalize = True))

False    0.624686
True     0.375314
Name: potential_acronym_in_url, dtype: float64
False    0.57974
True     0.42026
Name: potential_acronym_in_url_extra, dtype: float64


adding the extra possible acronyms increases the hit rate by another 12%. We can do some testing with an AUC-ROC model to figure out if we're having issues with false positives, but the limited training set is problematic.

In [13]:
def check_any_keyword(x):
    in_url = ([ ac in x['URL'] for ac in x['keywords'] ])
    return(any(in_url))

for df in dfs:
    df['keyword_in_url'] = df.apply(check_any_keyword, axis=1)

In [14]:
for df in dfs:
    df['matched_url'] = df['keyword_in_url'] | df['potential_acronym_in_url']
    df['matched_url_extra'] = df['keyword_in_url'] | df['potential_acronym_in_url_extra']
    
sample_df['correct_prediction'] = sample_df['isWrong'] == ~sample_df['matched_url']
sample_df['correct_prediction_extra'] = sample_df['isWrong'] == ~sample_df['matched_url_extra']

print(sample_df['correct_prediction'].value_counts())
print(sample_df['correct_prediction_extra'].value_counts())

True     95
False     5
Name: correct_prediction, dtype: int64
True     96
False     4
Name: correct_prediction_extra, dtype: int64


In [15]:
sample_df[sample_df['correct_prediction'] == False]

Institution                       URL  \
13                         Benedictine University               www.ben.edu   
42  New Mexico Institute of Mining and Technology               www.nmt.edu   
58               The General Theological Seminary  www.remingtoncollege.edu   
68                               Sunstate Academy  www.sandiegojobcorps.org   
94        Teachers College of Columbia University                www.tc.edu   

    isWrong      stripped_url                           keywords acronym  \
13        0               ben                      [benedictine]      bu   
42        0               nmt  [new, mexico, mining, technology]   nmimt   
58        1  remingtoncollege   [general, theological, seminary]     gts   
68        1  sandiegojobcorps                [sunstate, academy]      sa   
94        0                tc               [teachers, columbia]    tccu   

     potential_acronyms potential_acronym_in_url  \
13                 [bu]                    False   
42  [nmimt, mimt, nmim]                    False   
58        [gts, ts, gt]                     True   
68                 [sa]                     True   
94     [tccu, ccu, tcc]                    False   

             potential_acronyms_extra potential_acronym_in_url_extra  \
13                           [bu, bu]                          False   
42  [nmimt, mimt, nmim, nmiomat, nmi]                          False   
58                [gts, ts, gt, tgts]                           True   
68                           [sa, sa]                           True   
94        [tccu, ccu, tcc, tcocu, tc]                           True   

   correct_prediction_1 correct_prediction_2 keyword_in_url matched_url  \
13                False                False          False       False   
42                False                False          False       False   
58                False                False          False        True   
68                False                False          False        True   
94                False                 True          False       False   

   matched_url_extra correct_prediction correct_prediction_extra  
13             False              False                    False  
42             False              False                    False  
58              True              False                    False  
68              True              False                    False  
94              True              False                     True

Pretty cool - we already hit a 97% success rate on the trial data.

In [16]:
print(accred_df['matched_url'].value_counts(normalize = True))
print(accred_df['matched_url_extra'].value_counts(normalize = True))

True     0.865617
False    0.134383
Name: matched_url, dtype: float64
True     0.889802
False    0.110198
Name: matched_url_extra, dtype: float64


Performance on training data may exaggerate success - looking at much bigger database of accredited schools, fit rate is only 89% or so

In [17]:
accred_df[accred_df['matched_url_extra'] == False]

Institution  \
5                American University of Health Sciences   
7                             Adult Vocational Services   
9                           Fortis Institute - Scranton   
27                        Career Academy of Hair Design   
31       Ohio Center for Broadcasting - Colorado Campus   
34                     Ohio State School of Cosmetology   
36     Ben Franklin Career & Technical Education Center   
37         Monongalia County Technical Education Center   
38                   Putnam Career and Technical Center   
42                           Marinello School of Beauty   
44                        Everest College - Los Angeles   
45                           Everest College - Torrance   
55                         Everest Institute - San Jose   
56                         Everest Institute - Brighton   
57                          Everest Institute - Hayward   
64        Ohio Academy - A Paul Mitchell Partner School   
69                              Delmarva Beauty Academy   
75                              College of the Sequoias   
76                     Empire Beauty School - Peekskill   
93                      Kankakee Academy of Hair Design   
103                                      Bryman College   
110          Aviation Institute of Maintenance - Dallas   
112                                  American Institute   
119   Irell and Manella Graduate School of Biologica...   
128                     Apex Academy of Hair Design Inc   
143                      Brightwood College - Las Vegas   
144                Yeshiva Gedolah Imrei Yosef D'Spinka   
146    Living Arts College@School of Communication Arts   
155                                  Neumont University   
158         Universal Technology College of Puerto Rico   
168                     International Technical College   
171                    Northwest HVAC/R Training Center   
176                                     Medacom College   
184                           Maric College - San Diego   
186      House of Heavilin Beauty College - Kansas City   
187     House of Heavilin Beauty College - Blue Springs   
188          House of Heavilin Beauty College - Raymore   
196   Los Angeles County College of Nursing and Alli...   
217   Health Careers Center of Mercer County Technic...   
219                                      Fortis College   
220                           Fortis Institute - Towson   
235                               Monterey Adult School   
236                    Queen of the Holy Rosary College   
239                       Ner Israel Rabbinical College   
242   University of Nebraska - Nebraska College of T...   
248                   Akron School of Practical Nursing   
277                      Martinez School of Cosmetology   
279           South Central Louisiana Technical College   
283   American Museum of Natural History - Richard G...   
293   Ohio State School of Cosmetology & Experts Bar...   
297     Sherrill's University of Barber and Cosmetology   
298            South Branch Career and Technical Center   
305                             Southeastern University   
324              Lincoln College of Technology - Dayton   
326             Sylvia's International School of Beauty   
329              Ventura Adult and Continuing Education   
330                           Kaplan College - Columbus   
360                     Institute of Technology - Salem   
367                           Four Rivers Career Center   
381   West Side Montessori School's Teacher Educatio...   
388                    Nevada Regional Technical Center   
411                        Keystone Technical Institute   
425                                    Point University   
429                 Alliance City Schools Career Centre   
431   Sunnyvale - Cupertino Adult & Community Education   
434        Ashtabula County Technical and Career Campus   
442                    East-West Healing Arts Institute   
446                Academy for F

I feel pretty OK with practically all of the misses, except that it bothers me that my algorithm misses some fairly obvious matches like University of Wisconsin - Madison -> wisc and Washington College -> washcoll. I'm going to make one last variable that checks if potential abbreviations are included in the URL name.

In [18]:
def make_abbreviations(keywords):
    ## use the first three letters of each keyword as a potential abbreviation used in url
    return [ s[:3] for s in keywords ]

for df in dfs:
    df['potential_abbreviations'] = df['keywords'].apply(make_abbreviations)

def check_any_abbrev(x):
    in_url = ([ ab in x['stripped_url'] for ab in x['potential_abbreviations'] ])
    return(any(in_url))

for df in dfs:
    df['abbrev_in_url'] = df.apply(check_any_abbrev, axis=1)

In [19]:
accred_df

Institution  \
0                  American Commercial College of Texas   
1                 American Commercial College - Abilene   
2                          Anson College of Cosmetology   
3                American College of Medical Technology   
4     Aviation and Electronic Schools of America - C...   
5                American University of Health Sciences   
6                            Gerber Akron Beauty School   
7                             Adult Vocational Services   
8                  Alexandria Academy of Beauty Culture   
9                           Fortis Institute - Scranton   
10                    American Beauty Academy - Wheaton   
11       American College of Hairstyling - Cedar Rapids   
12                                      Antioch College   
13                      Antonelli College - Hattiesburg   
14               Academy of Oriental Medicine at Austin   
15                      Empire Beauty School - Glendale   
16                   Empire Beauty School - Westminster   
17                   Ashdown College of Health Sciences   
18    Ohio State University Agricultural Technical I...   
19                   ATI Career Training Center - Miami   
20                          Institute of Audio Research   
21                   Alaska Vocational Technical Center   
22                           Allegheny Wesleyan College   
23     Inter American University of Puerto Rico Bayamon   
24                                    Beacon University   
25                                         Beal College   
26                             Career Academy of Beauty   
27                        Career Academy of Hair Design   
28                                  Bellevue University   
29    Bellin College/ Bellin Health School of Radiol...   
30                             Bellingham Beauty School   
31       Ohio Center for Broadcasting - Colorado Campus   
32                                Berkeley City College   
33      Northwestern Institute of Health and Technology   
34                     Ohio State School of Cosmetology   
35                         Alderson Broaddus University   
36     Ben Franklin Career & Technical Education Center   
37         Monongalia County Technical Education Center   
38                   Putnam Career and Technical Center   
39                                  Bramson ORT College   
40                                  Baton Rouge College   
41                            Brewster Technical Center   
42                           Marinello School of Beauty   
43                            Brownson Technical School   
44                        Everest College - Los Angeles   
45                           Everest College - Torrance   
46                          Burlington Technical Center   
47                             California Career School   
48                                       Canada College   
49                        Capitol Technology University   
50                              Carnegie Career College   
51                              Carousel Beauty College   
52              Carousel of Miami Valley Beauty College   
53          Carousel Beauty College & The Spa Institute   
54    Cossatot Community College of the University o...   
55                         Everest Institute - San Jose   
56                         Everest Institute - Brighton   
57                          Everest Institute - Hayward   
58    Central California School of Continuing Educat...   
59                           Commercial Driver Training   
60                 Career Education Institute-Henderson   
61               Centro de Estudios Multidisciplinarios   
62                                Central Career School   
63               Dale Carnegie Training of Central Ohio   
64        Ohio Academy - A Paul Mitchell Partner School   
65           Center for Employment Training - Cleveland   
66              Central Florida Institute - Palm Harbor   
67                              

In [20]:
for df in dfs:
    df['matched_url_abbrev'] = df['keyword_in_url'] | df['potential_acronym_in_url_extra'] | df['abbrev_in_url']
    
sample_df['correct_prediction_abbrev'] = sample_df['isWrong'] == ~sample_df['matched_url_abbrev']

print(sample_df['correct_prediction'].value_counts())
print(sample_df['correct_prediction_extra'].value_counts())
print(sample_df['correct_prediction_abbrev'].value_counts())

True     95
False     5
Name: correct_prediction, dtype: int64
True     96
False     4
Name: correct_prediction_extra, dtype: int64
True     96
False     4
Name: correct_prediction_abbrev, dtype: int64


In [21]:
sample_df[sample_df['correct_prediction_abbrev'] == False]

Institution                       URL  \
39                   Newbridge College - El Cajon          www.newpaltz.edu   
42  New Mexico Institute of Mining and Technology               www.nmt.edu   
58               The General Theological Seminary  www.remingtoncollege.edu   
68                               Sunstate Academy  www.sandiegojobcorps.org   

    isWrong      stripped_url                           keywords acronym  \
39        1          newpaltz             [newbridge, el, cajon]    ncec   
42        0               nmt  [new, mexico, mining, technology]   nmimt   
58        1  remingtoncollege   [general, theological, seminary]     gts   
68        1  sandiegojobcorps                [sunstate, academy]      sa   

     potential_acronyms potential_acronym_in_url  \
39     [ncec, cec, nce]                    False   
42  [nmimt, mimt, nmim]                    False   
58        [gts, ts, gt]                     True   
68                 [sa]                     True   

             potential_acronyms_extra potential_acronym_in_url_extra  \
39         [ncec, cec, nce, ncec, nc]                          False   
42  [nmimt, mimt, nmim, nmiomat, nmi]                          False   
58                [gts, ts, gt, tgts]                           True   
68                           [sa, sa]                           True   

              ...            correct_prediction_2 keyword_in_url matched_url  \
39            ...                            True          False       False   
42            ...                           False          False       False   
58            ...                           False          False        True   
68            ...                           False          False        True   

   matched_url_extra correct_prediction correct_prediction_extra  \
39             False               True                     True   
42             False              False                    False   
58              True              False                    False   
68              True              False                    False   

   potential_abbreviations abbrev_in_url matched_url_abbrev  \
39          [new, el, caj]          True               True   
42    [new, mex, min, tec]         False              False   
58         [gen, the, sem]         False               True   
68              [sun, aca]         False               True   

   correct_prediction_abbrev  
39                     False  
42                     False  
58                     False  
68                     False  

[4 rows x 21 columns]

pretty interesting - with abbreviations, all of a sudden Newbridge College - El Cajon could reasonably have a URL of newpaltz.edu since new shared by both. Abbreviations may also add to the problem of false matching by guessing on too many potential acronyms.

In [24]:
print(accred_df['matched_url'].value_counts(normalize = True))
print(accred_df['matched_url_extra'].value_counts(normalize = True))
print(accred_df['matched_url_abbrev'].value_counts(normalize = True))

True     0.865617
False    0.134383
Name: matched_url, dtype: float64
True     0.889802
False    0.110198
Name: matched_url_extra, dtype: float64
True     0.913301
False    0.086699
Name: matched_url_abbrev, dtype: float64


Nice to see that adding the abbreviations in paid off - bump to 91.3% success rate with accredited institutions is pretty good. From visual inspection many of the remaining misses would require a great deal of extra effort to account for (e.g. Aviation Institute of Maintenance - Dallas -> fixthatplane.com)

In [25]:
accred_df[accred_df['matched_url_abbrev'] == False]

Institution  \
5                American University of Health Sciences   
7                             Adult Vocational Services   
9                           Fortis Institute - Scranton   
27                        Career Academy of Hair Design   
31       Ohio Center for Broadcasting - Colorado Campus   
34                     Ohio State School of Cosmetology   
36     Ben Franklin Career & Technical Education Center   
42                           Marinello School of Beauty   
44                        Everest College - Los Angeles   
45                           Everest College - Torrance   
55                         Everest Institute - San Jose   
56                         Everest Institute - Brighton   
57                          Everest Institute - Hayward   
64        Ohio Academy - A Paul Mitchell Partner School   
69                              Delmarva Beauty Academy   
75                              College of the Sequoias   
76                     Empire Beauty School - Peekskill   
93                      Kankakee Academy of Hair Design   
103                                      Bryman College   
110          Aviation Institute of Maintenance - Dallas   
112                                  American Institute   
128                     Apex Academy of Hair Design Inc   
143                      Brightwood College - Las Vegas   
144                Yeshiva Gedolah Imrei Yosef D'Spinka   
146    Living Arts College@School of Communication Arts   
155                                  Neumont University   
176                                     Medacom College   
184                           Maric College - San Diego   
186      House of Heavilin Beauty College - Kansas City   
187     House of Heavilin Beauty College - Blue Springs   
188          House of Heavilin Beauty College - Raymore   
196   Los Angeles County College of Nursing and Alli...   
219                                      Fortis College   
220                           Fortis Institute - Towson   
235                               Monterey Adult School   
236                    Queen of the Holy Rosary College   
239                       Ner Israel Rabbinical College   
242   University of Nebraska - Nebraska College of T...   
248                   Akron School of Practical Nursing   
277                      Martinez School of Cosmetology   
279           South Central Louisiana Technical College   
283   American Museum of Natural History - Richard G...   
293   Ohio State School of Cosmetology & Experts Bar...   
297     Sherrill's University of Barber and Cosmetology   
305                             Southeastern University   
324              Lincoln College of Technology - Dayton   
329              Ventura Adult and Continuing Education   
330                           Kaplan College - Columbus   
360                     Institute of Technology - Salem   
367                           Four Rivers Career Center   
381   West Side Montessori School's Teacher Educatio...   
388                    Nevada Regional Technical Center   
411                        Keystone Technical Institute   
425                                    Point University   
429                 Alliance City Schools Career Centre   
431   Sunnyvale - Cupertino Adult & Community Education   
434        Ashtabula County Technical and Career Campus   
442                    East-West Healing Arts Institute   
448                                         ATI College   
456                       Bilingual Education Institute   
459   Association Free Lutheran Theological Seminary...   
472                        American National University   
498                             Northwest Vista College   
499                                   Palo Alto College   
501                                St. Philip's College   
529                       Fortis Institute - Forty Fort   
562             Lincoln Technical Institute - Fern Park   
565                             

Probably the next step would be to feed in names of major state university systems (e.g. SUNY) or state abbreviations as something to look out for, but that's beyond what's needed for the current problem.